[sdl2-mixer] Bump to 2.6.3

FtZPetruska opened the pull request:
- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)
- [X] SHA512s are updated for each updated download
- [X] The "supports" clause reflects platforms that may be fixed by this new version
- [X] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [X] Any patches that are no longer applied are deleted from the port's directory.
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [X] Only one version is added to each modified port's versions file.

FtZPetruska committed:
[sdl2-mixer] Bump to version 2.6.3

MonicaLiu0311 commented:
Test feature `FluidSynth` failed on `x64-windows-static`.
```
PS E:\sdl2-mixer> ./vcpkg install sdl2-mixer[FluidSynth]:x64-windows-static
Computing installation plan...
The following packages will be built and installed:
    sdl2-mixer[core,fluidsynth,nativemidi]:x64-windows-static -> 2.6.3
Detecting compiler hash for triplet x64-windows-static...
Restored 0 package(s) from C:\Users\monica\AppData\Local\vcpkg\archives in 270.5 us. Use --debug to see more details.
Installing 1/1 sdl2-mixer:x64-windows-static...
Building sdl2-mixer[core,fluidsynth,nativemidi]:x64-windows-static...
-- Using cached libsdl-org-SDL_mixer-release-2.6.3.tar.gz.
-- Cleaning sources at E:/sdl2-mixer/buildtrees/sdl2-mixer/src/ease-2.6.3-13d61e43d0.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source E:/sdl2-mixer/downloads/libsdl-org-SDL_mixer-release-2.6.3.tar.gz
-- Applying patch fix-pkg-prefix.patch
-- Applying patch fix-pkgconfig.patch
-- Using source at E:/sdl2-mixer/buildtrees/sdl2-mixer/src/ease-2.6.3-13d61e43d0.clean
-- Found external ninja('1.11.0').
-- Configuring x64-windows-static
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
    Command failed: "C:/Program Files/Microsoft Visual Studio/2022/Enterprise/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe" -v
    Working Directory: E:/sdl2-mixer/buildtrees/sdl2-mixer/x64-windows-static-rel/vcpkg-parallel-configure
    Error code: 1
    See logs for more information:
      E:\sdl2-mixer\buildtrees\sdl2-mixer\config-x64-windows-static-dbg-CMakeCache.txt.log
      E:\sdl2-mixer\buildtrees\sdl2-mixer\config-x64-windows-static-rel-CMakeCache.txt.log
      E:\sdl2-mixer\buildtrees\sdl2-mixer\config-x64-windows-static-out.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:248 (vcpkg_execute_required_process)
  ports/sdl2-mixer/portfile.cmake:22 (vcpkg_cmake_configure)
  scripts/ports.cmake:147 (include)
```
**config-x64-windows-static-out.log:**
```
CMake Error at E:/sdl2-mixer/downloads/tools/cmake-3.25.1-windows/cmake-3.25.1-windows-i386/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find FluidSynth (missing: FluidSynth_LIBRARY)
Call Stack (most recent call first):
  E:/sdl2-mixer/downloads/tools/cmake-3.25.1-windows/cmake-3.25.1-windows-i386/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindFluidSynth.cmake:17 (find_package_handle_standard_args)
  E:/sdl2-mixer/scripts/buildsystems/vcpkg.cmake:843 (_find_package)
  CMakeLists.txt:675 (find_package)
```

FtZPetruska committed:
[sdl2-mixer] Fix findFluidSynth on windows-static.

FtZPetruska committed:
[sdl2-mixer] vcpkg_minimum_required is not needed

FtZPetruska committed:
[sdl2-mixer] Control SDL2MIXER_MOD with libmodplug

If SDL2MIXER_MOD is left on when no backend is availble, CMake issues a
warning.

FtZPetruska committed:
[sdl2-mixer] Turn off SDL2MIXER_DEPS_SHARED

Similar issue to #28137. SDL2MIXER_OPUS_SHARED and
SDL2MIXER_VORBIS_VORBISFILE_SHARED need to be manually
turned off as they are unconditionally turned on.

FtZPetruska committed:
[sdl2-mixer] Set the correct vorbis backend.

Even when SDL2MIXER_VORBIS_VORBISFILE was set to ON, stb_vorbis would be
used as the default backend.

FtZPetruska committed:
[sdl2-mixer] Use vcpkg_install_copyright

FtZPetruska committed:
[sdl2-mixer] Ensure flac and mpg123 are selected.

Turn off header-only libraries when the features are selected.

FtZPetruska commented:
The fluidsynth issue seemed to already be present before this PR. Fortunately, it was a very small patch. It seems like on `windows-static` triplets, a `lib` prefix is added to fluidsynth's lib file as such breaking SDL_mixer's `findFluidSynth.cmake`.

Additionally, while fixing this problem, I found more issues with the other features with how they were linked, chosen at runtime, and loaded.

FtZPetruska commented:
It seems like there is now an issue with opusfile. The port does not install a pc file which SDL2_mixer expects to find. Moreover, there was another issue upstream with loading opusfile in `SDL2_mixerConfig.cmake` that has not been merged into the 2.6 branch (https://github.com/libsdl-org/SDL_mixer/pull/468).

As such I will set this PR to draft for now until these points are addressed.

FtZPetruska committed:
[sdl2-mixer] Patch missing call to find_dependency

This is fixed upstream but has not been included in the 2.6.x branch.

FtZPetruska committed:
[sdl2-mixer] Fix dependency linking.

- The FLAC target is an upstream patch.
- VCPKG's opusfile port is static-only, meaning Opus and Ogg needs to be
linked explicitly.

FtZPetruska committed:
[sdl2-mixer] Update baseline.

FtZPetruska commented:
I rebased the branch on the recently merged opusfile changes. Since the vcpkg port is static-only, linking libogg and opus explicitly was needed.

I also noticed a build failure on Windows with the FLAC feature, which was fixed [upstream](https://github.com/libsdl-org/SDL_mixer/commit/2fe088d3c86e580855d38afcaf5064a7a6f5b1a7) but not merged in this release.

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg into bump-sdl2-mixer-2.6.3

MonicaLiu0311 commented:
All features are tested successfully in the following triplet:
x86-windows
x64-windows
x64-windows-static

dg0yt commented on ports/sdl2-mixer/fix-missing-find-opusfile.patch:
```diff
@@ -0,0 +1,14 @@
+     endif()
+ 
++    if(SDL2MIXER_OPUS AND NOT SDL2MIXER_VENDORED AND NOT TARGET opusfile::opusfile)
++        find_dependency(opusfile)
```
Why does it look for `opusfile` in exported config, but for `Opus` during the build?

FtZPetruska replied:
On Windows only I would get build failures with the `x64-windows` triplet, as the link step would fail if Opus and Ogg were not linked in addition to opusfile (see [1abdad3](https://github.com/microsoft/vcpkg/pull/29487/commits/1abdad359d10ed7064835bbada9f9ce7b9ccc56a)). 
On the other hand, when linking SDL_mixer to a project, neither Opus or Ogg are needed.

dg0yt replied:
I assume that `Opus::Opus` is from the `Opus` package.
If this target is used during the build, it will be in the exported config at least for static builds. And that it takes a corresponding `find_dependency` in the exported config.

FtZPetruska replied:
Thank you for the clarification. I had already opened a similar issue upstream about issues with static linking, but the outcome was:
- Either let the user handle it by setting the `<lib>_LINK_LIBRARIES` variable to the required additional libraries.
- Generate find/config files that capture all the dependency information like conan does.

In this context:
- libflac, libvorbis, and fluidsynth (once updated to 2.3.1) relying on their config file would bring their dependencies in static builds; 
- libmodplug and mpg123 do not have additional dependencies, so no change is required for them;
- opusfile lacks a CMake config file, so its dependencies (Opus and Ogg) need to be pulled in manually.
Should a patch add an explicit `CONFIG` in the `find_dependency` calls ?

dg0yt replied:
Upstream's advise is quite general. But basically my comments are exactly targeted at solving the  "capture all the dependency information". 
- If you add a `find_package` to the build and use the imported target, then you will have to add a corresponding `find_dependency` to the exported config.
- But first I would like to understand why you added the `find_package`. There are two possible situations.
  - Upstream's find procedures do not sufficiently capture usage requirements of direct dependencies, but vcpkg added proper config. Then we want to modify the buildsystem, to fix build and export.
  - A dependency doesn't properly export its usage requirements. Then we want to modify that dependency's port, not this port.

Thanks for researching the dependencies. I looked at opusfile some hours ago. With regard to this port, I would suggest to update the port to a recent commit, marking it version `0.12+20221122` or similar.
This commit brings a CMake build system and (official!) exported config, plus some good fixes but otherwise no significant changes to the implementation:
https://github.com/xiph/opusfile/compare/v0.12...9d71834

FtZPetruska replied:
Thank you for your feedback!

In this case, the added `find_package` calls are a result from `opusfile` not exporting its usage requirements.
Thank you for your help researching a good commit to update the port to. I will open a pull request shortly.

dg0yt commented on ports/sdl2-mixer/fix-deps-link.patch:
```diff
@@ -0,0 +1,33 @@
+             list(APPEND INSTALL_EXTRA_TARGETS ogg)
+         endif()
++    else()
++        find_package(Ogg CONFIG REQUIRED)
```
No corresponding `find_dependency` in exported config?

FtZPetruska replied:
Same reason as above, it felt more appropriate to add the `find_package` for Ogg in the `SDL2MIXER_OGG` block rather than the `SDL2MIXER_OPUS` block. But I can understand if you think it would be more appropriate to group them together.

dg0yt replied:
I'm not concerned about the block. 
I assume that `Ogg::Ogg` is from the `Ogg` package.
If this target is used during the build, it will be in the exported config at least for static builds. And that it takes a corresponding find_dependency in the exported config.

dg0yt commented on ports/sdl2-mixer/portfile.cmake:
```diff
@@ -1,29 +1,48 @@
+        mpg123 SDL2MIXER_MP3_DRMP3
 )
 
+if(FEATURE_OPTIONS MATCHES "SDL2MIXER_VORBIS_VORBISFILE=ON")
```
Don't depend on `vcpkg_check_features` implementation details.
Either
~~~cmake
if("libvorbis" IN_LIST FEATURES)
~~~
or
~~~cmake
if(SDL2MIXER_VORBIS_VORBISFILE)
~~~
(an official side effect of `vcpkg_check_features`).

dg0yt replied:
But this looks a little bit like an undesired switch between alternatives.

FtZPetruska replied:
Ah, I understand, thank you for explaining the official way to test for this!

The `libvorbis` feature has been available for a while, but since SDL_mixer 2.6.0 vorbis support can no longer be turned off, defaulting to the header-only `stb_vorbis`. What could be done is making `libvorbis` a default rather than a feature.

From my personal experience, I have had issues with the `stb_vorbis` backend in SDL_mixer in the past, so I would be inclined to make `libvorbis` the default. What do you think?

Similarly, currently the header-only libraries `dr_flac` and `dr_mp3` are used as the default backend if, respectively, `libflac` and `mpg123` are not selected. Should FLAC and MP3 support be fully disabled if the features are not selected ?

dg0yt replied:
I don't have a specific answer for sdl2-mixer. I'm concerned about the user being surprised when the backend changes as a side effect of installing another port which request a new feature from sdl2-mixer. In particular when the quality of backends varies.

FtZPetruska replied:
I would tend to favour turning off backends when they are not selected instead of falling back to a header-only alternative. And in the case of the vorbis backend where it cannot be turned off, take a dependency on libvorbis.

Would you like me to open a PR that implement the proposed changes ? Or open a discussion/issue about it first ?
